File: sol/solution/0001-0100/0031. Next Permutation/Solution.kt
class Solution {
    fun nextPermutation(nums: IntArray): Unit {
        var i = nums.size - 2
        while (i >= 0 && nums[i] >= nums[i + 1]) i--
        if (i >= 0) {
            var j = nums.size - 1
            while (nums[j] <= nums[i]) j--
            nums[i] = nums[j].also { nums[j] = nums[i] }
        }
        nums.reverse(i + 1, nums.size)
    }
    private fun IntArray.reverse(l: Int, r: Int) {
        var left = l
        var right = r - 1
        while (left < right) {
            this[left] = this[right].also { this[right] = this[left] }
            left++; right--
        }
    }
}

File: sol/solution/0001-0100/0031. Next Permutation/Solution.scala
object Solution {
  def nextPermutation(nums: Array[Int]): Unit = {
    var i = nums.length - 2
    while (i >= 0 && nums(i) >= nums(i + 1)) i -= 1
    if (i >= 0) {
      var j = nums.length - 1
      while (nums(j) <= nums(i)) j -= 1
      val temp = nums(i); nums(i) = nums(j); nums(j) = temp
    }
    reverse(nums, i + 1, nums.length - 1)
  }
  private def reverse(nums: Array[Int], l: Int, r: Int): Unit = {
    var left = l; var right = r
    while (left < right) {
      val temp = nums(left); nums(left) = nums(right); nums(right) = temp
      left += 1; right -= 1
    }
  }
}